src/toxcore/messenger/packet/msi.rs

```diff
@@ -0,0 +1,304 @@
+/*! Msi struct. Used by tox-av
+*/
+
+use nom::le_u8;
+
+use crate::toxcore::binary_io::*;
+
+/// Maximum size in bytes of msi message packet
+const MAX_MSI_PAYLOAD_SIZE: usize = 256;
+
+/*
+Msi Capabilities
+
+const MSI_CAP_S_AUDIO: u8 = 4;  /* sending audio */
+const MSI_CAP_S_VIDEO: u8 = 8;  /* sending video */
+const MSI_CAP_R_AUDIO: u8 = 16; /* receiving audio */
+const MSI_CAP_R_VIDEO: u8 = 32; /* receiving video */
+*/
+
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+/// Errors of msi session
+pub enum MsiErrorKind {
+    /// Error of none
+    MsiNone = 0,
+    /// Error of invalid message
+    InvalidMessage,
+    /// Error of invalid parameter
+    InvalidParam,
+    /// Error of invalid state
+    InvalidState,
+    /// Error of stray message
+    StrayMessage,
+    /// Error of system
+    System,
+    /// Error of handle
+    Handle,
+    /// Error of undisclosed
+    Undisclosed,
+}
+
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+/// Kind of msi request
+pub enum RequestKind {
+    /// Msi request of init
+    Init = 1,
+    /// Msi request of push
+    Push,
+    /// Msi request of pop
+    Pop,
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct Request(RequestKind);
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct MsiError(MsiErrorKind);
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct Capabilities(u8);
```

Comment:
Nope. Audio/AudioVideo/Video..

Reply:
```
const MSI_CAP_S_AUDIO: u8 = 4;  /* sending audio */
const MSI_CAP_S_VIDEO: u8 = 8;  /* sending video */
const MSI_CAP_R_AUDIO: u8 = 16; /* receiving audio */
const MSI_CAP_R_VIDEO: u8 = 32; /* receiving video */
```
It is possible someone has megaphone but don't have microphone.

Reply:
Anyway, I will revise it more friendly to human.

Reply:
I mean you must not keep it as a pure u8. You should parse these values:

```
const MSI_CAP_S_AUDIO: u8 = 4;  /* sending audio */
const MSI_CAP_S_VIDEO: u8 = 8;  /* sending video */
const MSI_CAP_R_AUDIO: u8 = 16; /* receiving audio */
const MSI_CAP_R_VIDEO: u8 = 32; /* receiving video */
```

And give `Capabilities` capabilities to tell the lib user, wherher you can send audio, send video... etc

Reply:
It seems we can use `bitflags` crate for this.

Reply:
Maybe

Reply:
Shall we use `bitflags`. I feel it is too much because we are not using bitwise OR or AND many times in our lib.

Reply:
But isn't it the place where we need it? Capabilities should be set independently to one byte using different bits. So it's the exact use case `bitflags` intends to solve.

Reply:
Ah. current FromBytes of CapabilitiesKind is not correct.
I can fix it using `bitflags` or raw coding. @kpp ?

Reply:
Try bitflags